### PR TITLE
Contango contracts

### PR DESCRIPTION
**Pull request by bbonanno**

### CODE REVIEW

File: contracts/contango/ContangoCauldron.sol [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    constructor(
+        uint128 _collateralisationRatio,
+        bytes6 _commonCurrency,
+        uint8 _commonCcyDecimals
+    ) {
+        collateralisationRatio = _collateralisationRatio * uint128(10**(18 - _commonCcyDecimals));
+        commonCcy = _commonCurrency;
+    }
+
+    function setParams(
```
Comment (alcueca) [inline contracts/contango/ContangoCauldron.sol:33]: I'd suggest splitting this function into `setRatio` and `setCurrency`.

File: contracts/contango/ContangoCauldron.sol [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    uint128 public collateralisationRatio;
+    bytes6 public commonCcy; // Currency to use as common ground for all the ink & art
+
+    constructor(
+        uint128 _collateralisationRatio,
+        bytes6 _commonCurrency,
+        uint8 _commonCcyDecimals
+    ) {
+        collateralisationRatio = _collateralisationRatio * uint128(10**(18 - _commonCcyDecimals));
```
Comment (alcueca) [inline contracts/contango/ContangoCauldron.sol:29]: Why not just passing the ratio?

File: contracts/contango/ContangoCauldron.sol [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+contract ContangoCauldron is Cauldron {
+    using CauldronMath for uint128;
+    using EnumerableSet for EnumerableSet.Bytes6Set;
+    using CastU256I256 for uint256;
+    using WMul for uint256;
+    using WDivUp for uint256;
+
+    EnumerableSet.Bytes6Set private assetsInUse;
+    mapping(bytes6 => DataTypes.Balances) public balancesPerAsset;
```
Comment (alcueca) [inline contracts/contango/ContangoCauldron.sol:18]: Are we doing away with individual vaults completely, or having a protocol vault and individual vaults in parallel?

File: contracts/contango/ContangoCauldron.sol [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        ) = vaultData(vaultId, true);
+
+        balances_ = _pour(vaultId, vault_, balances_, series_, ink, art);
+
+        _pour(vault_.ilkId, ink, vault_.seriesId, art);
+
+        return balances_;
+    }
+
+    function _pour(
```
Comment (alcueca) [inline contracts/contango/ContangoCauldron.sol:60]: I would do:
```
        bytes6 inkId,
        bytes6 artId,
        int128 ink,
        int128 art
```
All through the code, I've put identifiers first, amounts afterwards.

**Review by alcueca — COMMENTED** [review]
Mostly naming things, and one serious question about individual vaults.

File: contracts/contango/ContangoCauldron.sol (hunk 3 above)
Comment (bbonanno) [reply contracts/contango/ContangoCauldron.sol:18]: We have both, this impl calls the `_pour` on the parent that does all the existent stuff, then it replaces the call to `_level` by its internal call to `_pour` to update the global values and do global collateralisation check.
Maybe we could rename the second `_pour` to something else?

File: contracts/contango/ContangoCauldron.sol (hunk 1 above)
Comment (bbonanno) [reply contracts/contango/ContangoCauldron.sol:33]: ok

File: contracts/contango/ContangoCauldron.sol (hunk 2 above)
Comment (bbonanno) [reply contracts/contango/ContangoCauldron.sol:29]: no reason, started as storing the 3 things and ended up like this, will change it

File: contracts/contango/ContangoCauldron.sol (hunk 4 above)
Comment (bbonanno) [reply contracts/contango/ContangoCauldron.sol:60]: ok

File: contracts/contango/ContangoCauldron.sol [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        bytes6 artId,
+        int128 art
+    ) internal {
+        if (ink != 0) {
+            assetsInUse.add(inkId);
+            balancesPerAsset[inkId].ink = balancesPerAsset[inkId].ink.add(ink);
+        }
+        if (art != 0) {
+            assetsInUse.add(artId);
+            balancesPerAsset[artId].art = balancesPerAsset[artId].art.add(art);
```
Comment (alcueca) [inline contracts/contango/ContangoCauldron.sol:67]: It is a bit weird that all assets and series can be both collateral and debt. It's ok to have fyUSDC as collateral, but here you can have USDC as debt as well. You only call this from `pour` so you know that you are not doing that, but from a data integrity perspective it looks weak to me.

File: contracts/contango/ContangoLadle.sol [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+    constructor (ICauldron cauldron, IWETH9 weth) Ladle(cauldron, weth) { }
+
+    // @dev auth version of build, only contango can create vaults here
+    // all other methods rely on being vault owner, so no need to secure them
+    function build(bytes6 seriesId, bytes6 ilkId, uint8 salt)
+        external override payable auth
+        returns(bytes12, DataTypes.Vault memory)
+    {
+        return _build(seriesId, ilkId, salt);
```
[the remark this answers is not in the document]
Comment (alcueca) [reply contracts/contango/ContangoLadle.sol:16]: Now that `build` is `auth` they can be deterministic, no risk of DoS.

**Review by alcueca — CHANGES_REQUESTED** [review]
I think that we need to think again on the protocol vault, and code it more clearly.

File: contracts/contango/ContangoLadle.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (bbonanno) [reply contracts/contango/ContangoLadle.sol:16]: So I guess hardcoding a 0 to the internal call is what we want?

File: contracts/contango/ContangoCauldron.sol (hunk 5 above)
Comment (bbonanno) [reply contracts/contango/ContangoCauldron.sol:67]: All right, I'll rework that

File: contracts/contango/ContangoCauldron.sol [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        require(getFreeCollateralUSD() >= 0, "Undercollateralised");
+    }
+
+    function getFreeCollateralUSD() public returns (int256) {
+        (uint128 _collateralisationRatio, bytes6 _commonCurrency) = (collateralisationRatio, commonCcy);
+
+        uint256 totalInk;
+        uint256 totalArt;
+
+        //TODO is this reading .length() from storage each time?
```
Comment (alcueca) [inline contracts/contango/ContangoCauldron.sol:79]: As far as I understand, it is, @devtooligan?

File: contracts/contango/ContangoCauldron.sol [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        for (uint256 index; index < assetsInUse.length(); index++) {
+            bytes6 assetId = assetsInUse.get(index);
+            DataTypes.Balances memory _balances = balancesPerAsset[assetId];
+            if (assetId != _commonCurrency && (_balances.ink > 0 || _balances.art > 0)) {
```
Comment (alcueca) [inline contracts/contango/ContangoCauldron.sol:83]: You can rearrange this code:
```
if balances.art > 0
    totalArt += commonCurrency ? foo : bar
if balances.ink > 0 ...
```

File: contracts/contango/ContangoCauldron.sol [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+                IOracle oracle = spotOracles[assetId][_commonCurrency].oracle;
+                totalInk += _valueAsset(oracle, assetId, _commonCurrency, _balances.ink);
+                totalArt += _valueAsset(oracle, assetId, _commonCurrency, _balances.art);
+            } else {
+                //TODO maybe remove unsed asset?
```
Comment (alcueca) [inline contracts/contango/ContangoCauldron.sol:88]: Unused assets are ignored, so you can make that kind of garbage collection function an external one, called only when desired.

Comment (bbonanno) [reply contracts/contango/ContangoCauldron.sol:88]: yeah, was torn in between automatic vs manual :)
I guess manual it is then

File: contracts/contango/ContangoCauldron.sol (hunk 7 above)
Comment (devtooligan) [reply contracts/contango/ContangoCauldron.sol:79]: yes should prob cash the length to save gas.  

while we're at i think we can just cash the whole `assetsInUse` to save gas on read in line 81?

and can also do something about index++ --> ideally make the increment unchecked or at least use ++index instead

Comment (bbonanno) [reply contracts/contango/ContangoCauldron.sol:79]: ok, I'll do a pass on gas optimisations after I have the unit tests

Comment (bbonanno) [reply contracts/contango/ContangoCauldron.sol]: Can't cache the whole `assetsInUse` as it contains nested mappings, but I addressed the increment and cached the length

**Review by alcueca — APPROVED** [review]
Let's merge, great work.

**Pull request merged**